### PR TITLE
Bio1 2015 q3

### PR DESCRIPTION
**Pull request by AG-88301**

### CODE REVIEW

**Comment by vladimirfilip** [comment]
Good editorial. There's only a few typos to amend

File: solutions/bio1/ModernArt/index.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        <SText>
+            If the length of the string is $1$, the number of ways to contruct the string is the number of letters with a count greater than $0$.
+        </SText>
+      </SList>
+      <SText>
+        Finally, the number of ways to construct a string with a specified count of characters, is equal to the sum of:
+      </SText>
+      <SList>
+        <SText>
+            The number of ways to construct a string with length $length-1$ and $a-1$, $b$, $c$ and $d$ characters (make <SCode>"A"</SCode> the first character).
```
Comment (vladimirfilip) [inline solutions/bio1/ModernArt/index.tsx:37]: I feel like $len$ or $n$ would fit better in the sentence than $length$

File: solutions/bio1/ModernArt/index.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        <SText>
+            The number of ways to construct a string with length $length-1$ and $a$, $b$, $c-1$ and $d$ characters (make <SCode>"C"</SCode> the first character).
+        </SText>
+        <SText>
+            The number of ways to construct a string with length $length-1$ and $a$, $b$, $c$ and $d-1$ characters (make <SCode>"D"</SCode> the first character).
+        </SText>
+      </SList>
+      <SText>
+        We can count this recursively. To remove repeated work, we should memoize the results.
+        We can do this either by using a <SCode>std::map</SCode>, or Python's <SCode>functools.lru_cache</SCode>.
```
Comment (vladimirfilip) [inline solutions/bio1/ModernArt/index.tsx:51]: *using **_an_** $std::map$

**Review by vladimirfilip — CHANGES_REQUESTED** [review]

**Comment by macaquedev** [comment]
Sorry @AG-88301, but that won't cut it....
you've posted a solution that's twice as long as mine!

If you want you can just use mine, and then write that up I suppose?

`def get_num_perms(a, b, c, d):
    return fact[a + b + c + d] // (fact[a] * fact[b] * fact[c] * fact[d])


def solve(vals, n):
    if sum(vals) == 1:
        return chars[vals.index(1)]
    for curr in range(len(chars)):
        vals[curr] -= 1
        k = get_num_perms(*vals)
        if n <= k:
            return chars[curr] + solve(vals, n)
        else:
            n -= k
            vals[curr] += 1


if __name__ == "__main__":
    a, b, c, d, n = [int(x) for x in input("Enter 5 numbers: ").strip().split()]
    vals = [a, b, c, d]
    chars = "ABCD"
    fact = [1] * 50
    for i in range(2, 50):
        fact[i] = fact[i-1] * i

    print(solve(vals[:], n))`

**Comment by AG-88301** [comment]
> Sorry @AG-88301, but that won't cut it.... you've posted a solution that's twice as long as mine!
> 
> If you want you can just use mine, and then write that up I suppose?
> 
> ```
> def get_num_perms(a, b, c, d):
>     return fact[a + b + c + d] // (fact[a] * fact[b] * fact[c] * fact[d])
> 
> 
> def solve(vals, n):
>     if sum(vals) == 1:
>         return chars[vals.index(1)]
>     for curr in range(len(chars)):
>         vals[curr] -= 1
>         k = get_num_perms(*vals)
>         if n <= k:
>             return chars[curr] + solve(vals, n)
>         else:
>             n -= k
>             vals[curr] += 1
> 
> 
> if __name__ == "__main__":
>     a, b, c, d, n = [int(x) for x in input("Enter 5 numbers: ").strip().split()]
>     vals = [a, b, c, d]
>     chars = "ABCD"
>     fact = [1] * 50
>     for i in range(2, 50):
>         fact[i] = fact[i-1] * i
> 
>     print(solve(vals[:], n))
> ```

I can shorten mine if u want... i havent used any loops

**Comment by vladimirfilip** [comment]
I would't replace @AG-88301's solution with this. This could be presented as an alternative non-recursive method.

**Comment by vladimirfilip** [comment]
Fair enough, I was referring specifically to the mathematical method of counting combinations. I think that would be useful to include in the editorial

**Comment by vladimirfilip** [comment]
It's not a must, however.

**Review by vladimirfilip — DISMISSED** [review]

**Comment by AG-88301** [comment]
what review did i dismiss

**Review by macaquedev — APPROVED** [review]

**Pull request merged**